### DISCUSSION TITLE
Construction Menu rework

### DISCUSSION THREAD

**Discussion by karst** [OpenRCT2/OpenRCT2#17572] [Feature requests]
It's likely we will need a slight overhaul of the construction menu to implement new features. 

@X123M3-256 wants to add catwalks at some point as toggle able addition

@spacek531 wants to add togglable brakes, I believe.

All these things need a place on the construction menu.

Here's a mockup for what it could look like and some ideas what could change.

![image](https://user-images.githubusercontent.com/1516893/179870799-4563c816-8d40-4787-9972-7417b459552a.png)

This is the mockup I have made thusfar. 

Ideas:
- Remove special elements menu (to clear space) and replace for an icon (looping icon under track style, might need to be renamed to track options).
- Split Roll / Banking with Track styles.
- Track styles: I have currently laid this out next to each other, but I think, since they need to switch between anyhow, it might be smarter to have them toggle between in a drop down menu, like you would do with paths.
- Move chain lift to track options. 
- Permanent placement for brake, booster and chain lift speed + 4D rotation. (greys out when no element that can do it is selected)
- Add icons for special elements, brakes, boosters, catwalks and track styles.

With the mockup I tried keeping the window size about the same, while still adding more functionality.

I don't know if this is the right approach, it's why I am posting this discussion.

**Comment by ocalhoun6** [comment]
It _is_ starting to look a bit cluttered -- and that's coming from somebody who usually prefers to have as many buttons as possible. I could see how this window might look very intimidating to a new player who doesn't know what everything does.

Maybe there could be tabs? Like it defaults to the 'basic' or 'classic' tab that sticks very closely to the vanilla RCT2 build window, but then you can click on the 'advanced' tab to get all these new options on it... (And, of course, ideally, the game would remember which tab you'd selected most recently and then use that as the default for next time.)

**Reply by karst** [reply]
Ofcourse this is with all and everything enabled. Most rides won't use half of these buttons

**Reply by IntelOrca** [reply]
Have you looked at Locomotion? That has tabs on the build track window.

**Reply by oli414** [reply]
I like the idea of having tabs to keep things from becoming too visually crowded, in particular for the special track pieces

**Comment by spacek531** [comment]
I think it was X7123M3-256 who wanted brakes as a flag. I wanted additional banking angles: https://github.com/OpenRCT2/OpenRCT2/issues/17024#issuecomment-1105673493

And I wanted block brakes to have a speed setting, and I wanted brakes to have multidim rotation.

"One place" for rotation *and* speed may make sense right now, but it won't when a 4D darkride is added.

**Comment by duncanspumpkin** [comment]
We also need to add the ability to switch the track style so that people can do merges without having to use an additional track.

**Reply by ocalhoun6** [reply]
Hm... Maybe while we're at it, a toggle for: visible everything/invisible supports/invisible supports and invisible track.

That might save a lot of effort in the tile inspector later for people who want parts of the track to be invisible or have no supports.

**Reply by spacek531** [reply]
I believe this is represented by the bi-tri-quad track icons.

**Reply by karst** [reply]
Hmm, how it'd be sick to have an icon (visual representation) for every track style then these get enabled in the drop down menu when enabled with cheats

**Reply by X123M3-256** [reply]
I think this would be best done with a dropdown at the bottom like the arbitrary ride type menu. I think the icons are best for track types that are *supposed* to have multiple styles available, and there would generally only be two or three in that case.

**Comment by X123M3-256** [comment]
I think doing this now is premature, a lot of these suggested features won't be fully implemented for a while, and if we leave space for everything we might add in future then the menu now is going to look a bit empty now. I think we should just add to the existing ride construction menu until it becomes too cumbersome, and then think about how to rearrange it if necessary.

**Comment by X123M3-256** [comment]
I've also had a look at the track tile element and there's very little free space for new flags, so I don't think things like turning brakes and boosters into flags are going to be easy, let alone adding catwalks like I wanted to do. It looks like there's only *one* free track flag.

I can only see two potential workarounds. One is to use the 6 bits used for the door state/seat rotation, as these are only needed for specific ride types (of course, that means those ride types that need them can't also have these extra flags). The only other idea I have is to take the high order bits of the ride or ride type index (I mean, do we really need to support 65536 different ride types?).

**Reply by karst** [reply]
I wonder how much it would do to reduce it to 32768

Chainlift, booster and brake could be on the same flag, being 0 for disabled, 1 for chainlift, 2 for booster, 3 for brake orso perhaps?

**Reply by X123M3-256** [reply]
> I wonder how much it would do to reduce it to 32768

That gains you one bit.

Honestly, I really wish they hadn't added the ride type field. It uses up two whole bytes of space to support a hack that can be accomplished in multiple other ways, and leaves little space for anything else. There'd be no shortage of room for new flags and stuff if that field wasn't there. I feel like we could safely use the top 4 bits as flags which would limit the number of ride types to 4096, and I can't see a need for more than that.

> Chainlift, booster and brake could be on the same flag, being 0 for disabled, 1 for chainlift, 2 for booster, 3 for brake orso perhaps?

Which would require two flags, not one (I guess that's better than 3?)

I'd be inclined to keep brakes and boosters as track pieces simply because they can be (even if the UI is modified to presents them as a toggle), and implement different track styles as different ride types (I mean, they've allocated so much space for the ride type we might as well use it), which leaves whatever space we might be able to scrape together for new flags for things that cannot easily be implemented any other way.

Another option is to use the isCableLift flag - for rides which don't have a cable lift, this is unused, and could be used in conjunction with the lift hill flag to signify brake/booster pieces. Only problem is that rides which do have cable lifts can't do this.

**Reply by ocalhoun6** [reply]
Maybe things like a flag for catwalks should wait until after a new track design format?

That's the limitation here, right? And I thought we were working on a new track design format anyway.

**Reply by X123M3-256** [reply]
> Maybe things like a flag for catwalks should wait until after a new track design format?
> 
> That's the limitation here, right? And I thought we were working on a new track design format anyway.

No, the track design format is not the limitation, we're talking about the tile element structures (i.e the .park format), not the TD6 format (which wouldn't support it either but that's not the problem).

**Comment by spacek531** [comment]
I want that last track flag for supports visible

**Reply by X123M3-256** [reply]
But there's already a way to make supports invisible

**Reply by spacek531** [reply]
What method?

**Reply by X123M3-256** [reply]
> What method?

Move the support underneath the surface tile in the tile inspector

**Reply by spacek531** [reply]
That's an ugly hack and doesn't make supports visible on every straight track as is done in LL.

**Reply by X123M3-256** [reply]
> That's an ugly hack and doesn't make supports visible on every straight track as is done in LL.

Why would you want to draw supports on every tile? Surely the point of having a "remove support" flag is so you can get rid of the supports and build custom ones? I think the tiles on which default supports are drawn should be linked to the ride type as it is now.

**Reply by spacek531** [reply]
> Why would you want to draw supports on every tile?
Because RCT1 does it

> a "remove support" flag
The flag I am referring to is a "draw supports" flag

>  the tiles on which default supports are drawn should be linked to the ride type as it is now.
This is all implemented in my PR: https://github.com/OpenRCT2/OpenRCT2/pull/15356

**Reply by X123M3-256** [reply]
> Because RCT1 does it

If you just want to display RCT1 parks more accurately you could have it as a toggle like the transparent water. I can't see any other reason this would be desirable, since anyone wanting that level of control over where supports draw would probably be using custom supports anyway.

**Reply by spacek531** [reply]
This method acts as a support hider and RCT1 support in a single bit and it doesn't involve changing the tile order, so I think it has merit.

**Reply by X123M3-256** [reply]
> This method acts as a support hider and RCT1 support in a single bit and it doesn't involve changing the tile order, so I think it has merit.

I mean, I think it would be fine if this wasn't literally the only bit we have left. But everything I wanted to implement needs multiple free bits. I was thinking that maybe that bit could be used as a flag that says that some of the other bits should be interpreted differently, which is a hack but I don't have many other ideas.

**Comment by karst** [comment]
I was thinking of a useful feature, which would be that you can disconnect your building (broken chain icon?) and start somewhere else (as if you were starting a new ride, but it's still the same ride). And when you have it disconnected you can start from any track element. Would be especially be useful for trackitecture, custom supports, etc, but also for making rides similar to pantheon (mainly if we get something like track switch elements implemented).

**Reply by IntelOrca** [reply]
I am sure that is a heavily desired feature and probably would not be too hard to implement. We just need a button to change it back to the original place first track piece mode.

**Reply by X123M3-256** [reply]
> I was thinking of a useful feature, which would be that you can disconnect your building (broken chain icon?) and start somewhere else (as if you were starting a new ride, but it's still the same ride).

I think such a feature wouldn't just be useful for hacks. You'd be able to place an element where you want it and then build the connecting track after, instead of having to use trial and error to get the element positioned how you want it. This would be particularly useful for interlocking elements, or where you're not quite sure where ab element would fit.

**Reply by karst** [reply]
![image](https://user-images.githubusercontent.com/1516893/180824522-8d2e55e8-4e22-4618-9828-2b72f2174c4e.png)

Here's a sketch I made. Perhaps this could be a correct tool for it. Could be placed on the other side of the blue flag (so directly above the entrance button).

**Reply by Broxzier** [reply]
Just found a funny bug that already seems to partly make this possible, allowing you to place multiple starts:

![Forest Frontiers 2022-07-25 18-55-00](https://user-images.githubusercontent.com/9705046/180833231-92530f94-6139-4c6f-a76b-ed6005d21699.png)

So this is definitely easy to implement, just need a way to switch between modus manually.

**Reply by ocalhoun6** [reply]
This is actually quite similar to the two path building modes -- one that continues on from the last path piece placed, one that simply drops a path tile wherever you click. The only difference is that when building tracks, you have a lot more different options of what gets placed when you click.

So I think it's best to design the gui for this mode with the gui for the path building modes in mind, taking inspiration from that. This will help keep things more consistent, and consistency will help players more intuitively grasp what this new build mode does.
![Leafy Lake 2022-07-25 10-47-08](https://user-images.githubusercontent.com/26573303/180841544-dd66a723-28d7-46fe-8a4a-d5ffeca8d208.png)

**Reply by karst** [reply]
I don't think you want to be able to freely smear down track. Smearing also wouldnt automatically connect track, i personally think you'd overcomplicate the system if you make it act like paths. 

Here's an updated image for the tool. Idk if it's the icon we're looking for though.
![Snippet](https://user-images.githubusercontent.com/1516893/180841639-9f3b6887-a463-47b1-a252-4413db750e1a.png)

**Reply by ocalhoun6** [reply]
It wouldn't work _exactly_ like the path laying mode, of course. No smearing. Only one track piece per click. And preserving the ability to hold down shift and raise or lower it.

But I do think it's similar enough in concept to warrant being similar in interface.

**Reply by oli414** [reply]
One argument I'd like to make against this idea is that tracks are directional, and a new player will inevitably try to connect track pieces that run in different directions.

I still think this would be a valuable feature to have, just thinking if there's a way to solve that particular scenario

**Reply by karst** [reply]
would it be possible to have arrows pointing outwards the ends and into beginning of track while building? That way you don't have that confusion

**Reply by karst** [reply]
> But I do think it's similar enough in concept to warrant being similar in interface.

I don't think it's necessary. Having a cut off tool would be enough and other than that it would function the same as it always has. No need to make a new interface for it.

**Reply by karst** [reply]
![image](https://user-images.githubusercontent.com/1516893/180851721-9dbb9ce3-b7a0-4b01-87b8-a1771e1e5616.png)

Locomotion uses this. Could steal that.

I have also edited the thingy i was working on a bit more. It all depends on what you guys are looking for.
![Snippet](https://user-images.githubusercontent.com/1516893/180851808-58587928-fceb-47db-a944-4a8840a10180.png)

**Reply by karst** [reply]
![Pointer](https://user-images.githubusercontent.com/1516893/180856471-f23d2f6f-8f0e-43bd-b5c5-34649c97c0db.png)


Here's one that resembles an OpenRCT2 version of what you'd find in Locomotion for the same function.

**Reply by karst** [reply]
> One argument I'd like to make against this idea is that tracks are directional, and a new player will inevitably try to connect track pieces that run in different directions.
> 
> I still think this would be a valuable feature to have, just thinking if there's a way to solve that particular scenario

![image](https://user-images.githubusercontent.com/1516893/182504538-a3661e6d-30aa-4f8f-bb26-12d9e8ce9d69.png)
this is how Loco seems to solve that issue

**Reply by spacek531** [reply]
The way Loco solves the issue is by programming the game to allow trains to travel on tracks backwards.

**Reply by karst** [reply]
That'd be even better ofcourse

**Reply by 733737** [reply]
> i agree, having a junior coaster that leaves station up a lift hill in the back then keeps going around a circuit would be fun

**Reply by ocalhoun6** [reply]
>allow trains to travel on tracks backwards.

Could potentially have weird effects on special/directional track pieces, though?

Especially, going into a station piece the wrong way around might do weird things. (Then again, it might end up being a positive, once players find new and interesting ways to use that glitch to create hacked rides.)

**Reply by 733737** [reply]
station pieces would need to be 2way obviously, there is no way the existing station code would remain intact

**Comment by Basssiiie** [comment]
With the special elements dropdown currently in the game you can see which element you currently have selected. How does this work with the looping icon? Will there be icons for all the special elements? E.g. barrel rolls, corkscrews, S-bends, photo sections, vertical drop brake, river rapid elements, wild mouse rotation control, monster truck logs...?

**Reply by karst** [reply]
That's a good question, maybe it might be smartest to have the name of the object display in the top center of the preview.

**Reply by X123M3-256** [reply]
I really don't see what's wrong with the dropdown.

**Reply by karst** [reply]
Could clear some extra space, since in this case a new area is added making the menu larger. Could use that for a singular icon which takes up less space in the end.

**Comment by 733737** [comment]
how's this? 
![compactridewindow](https://user-images.githubusercontent.com/94884086/181406850-25746b2a-921f-4b14-b5e5-9912d2a1a672.png)

**Reply by spacek531** [reply]
how does it look when you have a quarter loop?

**Reply by 733737** [reply]
I don't know, I just took the other image and made it more compact since I personally would like a more compact build window

**Comment by karst** [comment]
![Construction-Window](https://user-images.githubusercontent.com/1516893/198891968-ca31c716-a8d6-4bae-a7e4-9f8188983e0c.png)

I made a new mock up from scratch. I think this will be a lot more benificial than the previous mockup I made.

The crosshair is a feature from Loco I'd like to see in OpenRCT2 where you can easily disconnect and continue building somewhere else (snippet tool). When you to place down the first piece or snippet piece this would be a rotate button instead.

The track button would be arbitrary track type changes. This is a drop down menu. By default this only includes a list of supported tracks, but with cheats enabled it will show all available track types. For this a lot of icons need to be made, I'll gladly make (some of) these, I think the effort will be worth it. These track type icons already exist:
![image](https://user-images.githubusercontent.com/1516893/198892108-308a35f1-b3a8-49ee-9a13-e98fd129e4a9.png).

I took massive inspiration of RCT3, Planet Coaster and Parkitect, who all use a horizontal menu rather than a vertical one for construction.

Please do let me know if I forgot anything.

**Reply by ocalhoun6** [reply]
- Generally quite like it ... though I'm not sure that moving to a horizontal layout is necessary or a good idea. IMO, it should only deviate from the vanilla layout if it's really necessary to do so.
- It's good that there's space left over in the middle for potential future rides with multiple banking angles available. (For example, you could easily add icons for a 90 degree banking here.)
- Where'd you get the icons for the S-bends? (I assume that's what they are.) The bend in them seems too extreme, and would be better if they more closely resembled the actual shape of the s-bend track.
- I think you could make the second panel (from left to right) narrower if you do two things: A) move the ultra-wide curves and s-bends to the top row of icons, while moving the other curves to the bottom row -- this uses the space more efficiently. B) Dynamically change which of the slope icons are shown, since not all of them can be built at any one time anyway -- this lets you cut some horizontal space from that row.
- The entrance/exit buttons could be made into icons and moved up into the set of 6 icons just above them, which would make for a cleaner look IMO.
- The track type icon could be moved to the left, leaving nothing under the special elements menu ... which means the special elements menu could be a little bit taller and therefore less likely to need the scroll bar (depending on ride type).
- Not sure if the icons for brakes/booster/photo section/block brakes are really ideal. Took me a while to figure out what they were supposed to be.
- How would this handle a ride that might be setting brake/booster speed *and* seat rotation at the same time, such as a brake piece on a multi-dimension coaster?

I could make a mockup of those ideas/improvements, if that would help.

**Reply by karst** [reply]
I could imagine with #18478 this could be made without having to make it the default option, or perhaps as a togglable option for people that are more used to this layout since it's used by every other coaster builder. 
To respond to @ocalhoun6's questions
* The S-Bends are directly taken from OpenLoco's G1. I could try and changing things around but I don't think that'll be a very necessary change, especially to have coherency in graphics with Locomotion (not that this is necessary).
* I don't feel like changing up the second panel to much as that'd just clutter up the menu. Giving each appropriate space is fine.
* Making icons for entrance and exit was my intention if I end up making this. I don't feel it should be with lift and common special elements, that does not make sense in any way. Should be outside of them. Not to mention I want to leave space for @X123M3-256's implementation of (vertical) boosters if that gets submitted/merged.
* I could see the track type/covered piece buttons be beside the banking but again, the space is there under the special elements menu, why not use it, doesn't look out of place either.
* Icons are quick mockups, so definitely don't represent what they'd be. @spacek531 made some icons yesterday that I actually really like and could see used. 
![image](https://user-images.githubusercontent.com/1516893/210801357-cc526d73-d3db-4a8f-8dc6-acb899d59f3b.png)
* The game does currently not support brake/booster speed and seat rotation at the same time, they share the same parameters afaik, though don't quote me on that.

**Reply by ocalhoun6** [reply]
Now that I'm coming back to it, I'd actually have one more suggestion:

Currently, you have the block that displays the current selected item (and the button to build it) on the far left of the window. But the ride construction window typically opens in the top left of the screen, which means this placement will put the most frequently used button as far as possible from the center of the screen when working.

I think it would be more efficient/ergonomic to move this element to the far right section of your window, moving the special elements block to the far left instead. That way, you don't have to move the mouse as far in order to build the selected track piece. (I'd probably go for mirroring the whole layout, really. Because the curves/slopes/banking will also be used more often than the stations/brakes/etc block.)

>The game does currently not support brake/booster speed and seat rotation at the same time, they share the same parameters afaik, though don't quote me on that.

Currently, yes ... though I think there are plans to eventually decouple those so that (for example) seat rotation can still be set on brake pieces. Might as well have the UI already capable of supporting both at once.

**Reply by spacek531** [reply]
> * The game does currently not support brake/booster speed and seat rotation at the same time, they share the same parameters afaik, though don't quote me on that.

OpenRCT2 does support simultaneous brake and seat rotation but TD6 does not.

> Currently, yes ... though I think there are plans to eventually decouple those so that (for example) seat rotation can still be set on brake pieces. Might as well have the UI already capable of supporting both at once.

That will be part of speedy block brakes, but currently only on block brakes and not brakes for TD6 reasons.

**Comment by karst** [comment]
Implementation of a reworked build menu which will feel familiar in the hands of our users, but allow for new features is of importance. 
I think having the different orientations available as options would be great, but it means redesigning two uis. I do not mind that myself. 
I think the smartest thing to add to this is modularity. If a new track element requires more options a side window which will be attached to the window will pop out to allow for setting parameters.
For example: when selecting brakes on the 4th Dimension coaster it will pop out a side-window to allow for seat rotation (or speed, whatever works there that doesn't really matter). Or for implementation of extra booster options a side window can pop up that allows you to select the forewards and backwards modes.
For flat to steep a side window would pop up that allows for transition sizes, this can be a slider or a menu with different icons.
Switching between track styles could work the same way.
This modularity would also mean a not so drastic change is necessary on the UI departement and could even already be implemented.

This kind of modularity is already available in a different UI, which is the staff window. It opens a sidewindow where you can select the patrol area's size.

**Reply by SpartanFrederic104** [reply]
On the topic of switching between track styles, would it also be possible to add an option to switch between support structure styles?

Additionally, I would love to see an option to build track pieces with anny one of the 3 alternate color schemes selected instead of the standard colors for the ride.

**Reply by X123M3-256** [reply]
> On the topic of switching between track styles, would it also be possible to add an option to switch between support structure styles?

Kind of ... the supports are tied to the track type but you could have two selectable track styles that differ only in their supports. I think being able to arbitrarily mix and match any track and support style isn't feasible though. There's very little free space left in the tile element ... not really anywhere to store a support type as well as a track type, and if there was space I can think of lots of other potential uses for it.

I don't think there are a lot of uses for this but Arrow rides could be one, they often use a mix of truss and cylindrical column supports.

**Comment by J-PIE-314** [comment]
Issue #22933 is relevant to this conversation.

**Reply by victwnd** [reply]
edited the mockup posted above (not coloured cause this was just done on a whim) to include more buttons in the slope section (and tighter button placement in the direction section to possibly fit a tiny turns button for a coaster that would use all turn types?)

the question mark would have the large steep to horizontal piece sprite, but i have no idea how to draw that or where to put it tbh
![image](https://github.com/user-attachments/assets/ce3db8d2-2464-431a-a3db-2a33fe3e67ca)

**Comment by J-PIE-314** [comment]
This request is related to https://github.com/OpenRCT2/OpenRCT2/discussions/17583 and https://github.com/OpenRCT2/OpenRCT2/discussions/18392.
